include/crow/websocket.h

```diff
@@ -118,18 +118,48 @@ namespace crow
                 start(crow::utility::base64encode((unsigned char*)digest, 20));
             }
 
+            ~Connection() noexcept override
+            {
+                // Do not modify anchor_ here since writing shared_ptr is not atomic.
+                auto watch = std::weak_ptr<void>{anchor_};
+
+                // Wait until all unhandled asynchronous operations to join.
+                // As the deletion occurs inside 'check_destroy()', which already locks
+                //  anchor, use count can be 1 on valid deletion context.
+                while (watch.use_count() > 2) // 1 for 'check_destroy() routine', 1 for 'this->anchor_'
+                {
+                    std::this_thread::yield();
+                }
+            }
+
             /// Send data through the socket.
             template<typename CompletionHandler>
-            void dispatch(CompletionHandler handler)
+            void dispatch(CompletionHandler&& handler)
             {
-                adaptor_.get_io_service().dispatch(handler);
+                adaptor_.get_io_service().dispatch(
+                  [this,
+                   handler = std::forward<CompletionHandler>(handler),
+                   watch = std::weak_ptr<void>{anchor_}] {
+                      auto anchor = watch.lock();
+                      if (anchor == nullptr) { return; }
+
+                      handler();
+                  });
             }
 
             /// Send data through the socket and return immediately.
             template<typename CompletionHandler>
-            void post(CompletionHandler handler)
+            void post(CompletionHandler&& handler)
             {
-                adaptor_.get_io_service().post(handler);
+                adaptor_.get_io_service().post(
+                  [this,
+                   handler = std::forward<CompletionHandler>(handler),
+                   watch = std::weak_ptr<void>{anchor_}] {
+                      auto anchor = watch.lock();
+                      if (anchor == nullptr) { return; }
+
+                      handler();
+                  });
```

Comment:
Almost forgot when testing - c++11 😓 .

I guess you can make a custom functor with this behaviour that will wrap the handler. But please, don't add over engineered code...